pinot-broker/src/main/java/org/apache/pinot/broker/cursors/FsResponseStore.java

```diff
@@ -0,0 +1,211 @@
+    List<String> requestIdList = new ArrayList<>(queryPaths.size());
+
+    LOGGER.debug(String.format("Found %d paths.", queryPaths.size()));
+
+    for (FileMetadata metadata : queryPaths) {
+      LOGGER.debug(String.format("Processing query path: %s", metadata.toString()));
+      if (metadata.isDirectory()) {
+        try {
+          URI queryDir = new URI(metadata.getFilePath());
+          URI metadataFile = combinePath(queryDir, String.format(RESPONSE_FILE_NAME_FORMAT, _fileExtension));
+          boolean metadataFileExists = pinotFS.exists(metadataFile);
+          LOGGER.debug(
+              String.format("Checking for query dir %s & metadata file: %s. Metadata file exists: %s", queryDir,
+                  metadataFile, metadataFileExists));
+          if (metadataFileExists) {
+            BrokerResponse response =
+                _responseSerde.deserialize(pinotFS.open(metadataFile), CursorResponseNative.class);
+            requestIdList.add(response.getRequestId());
+            LOGGER.debug("Added query store {}", queryDir);
+          }
+        } catch (Exception e) {
+          LOGGER.error("Error when processing {}", metadata, e);
+        }
+      }
+    }
+
+    return requestIdList;
+  }
+
+  @Override
+  protected boolean deleteResponseImpl(String requestId)
+      throws Exception {
+    PinotFS pinotFS = PinotFSFactory.create(_dataDir.getScheme());
+    URI queryDir = combinePath(_dataDir, requestId);
+    if (pinotFS.exists(queryDir)) {
+      pinotFS.delete(queryDir, true);
+      return true;
+    }
+    return false;
+  }
+
+  @Override
+  protected void writeResponse(String requestId, CursorResponse response)
+      throws Exception {
+    PinotFS pinotFS = PinotFSFactory.create(_dataDir.getScheme());
+    URI queryDir = combinePath(_dataDir, requestId);
+
+    // Create a directory for this query.
+    pinotFS.mkdir(queryDir);
+
+    Path tempResponseFile = getTempPath(_localTempDir, "response", requestId);
+    URI metadataFile = combinePath(queryDir, String.format(RESPONSE_FILE_NAME_FORMAT, _fileExtension));
+    try {
+      _responseSerde.serialize(response, Files.newOutputStream(tempResponseFile));
+      pinotFS.copyFromLocalFile(tempResponseFile.toFile(), metadataFile);
+    } finally {
+      Files.delete(tempResponseFile);
+    }
+  }
+
+  @Override
+  protected long writeResultTable(String requestId, ResultTable resultTable)
+      throws Exception {
+    PinotFS pinotFS = PinotFSFactory.create(_dataDir.getScheme());
+    URI queryDir = combinePath(_dataDir, requestId);
+
+    // Create a directory for this query.
+    pinotFS.mkdir(queryDir);
+
+    Path tempResultTableFile = getTempPath(_localTempDir, "resultTable", requestId);
+    URI dataFile = combinePath(queryDir, String.format(RESULT_TABLE_FILE_NAME_FORMAT, _fileExtension));
+    try {
+      _responseSerde.serialize(resultTable, Files.newOutputStream(tempResultTableFile));
```

Comment:
I dug in deeper why this worked without a `close`. Seems like Jackson may or may not decide to close the stream. 

```
Method that can be used to serialize any Java value as JSON output, using output stream provided (using encoding JsonEncoding.UTF8.
Note: method does not close the underlying stream explicitly here; however, JsonFactor this mapper uses may choose to close the stream depending on its settings (by default, it will try to close it when JsonGenerator.
```

Ref: https://fasterxml.github.io/jackson-databind/javadoc/2.8/com/fasterxml/jackson/databind/ObjectWriter.html#writeValue(java.io.OutputStream,%20java.lang.Object)

Reply:
I will add a close.